Replaced removed stdlib crypt module with crypt-r
Index: waagent/requirements.txt
===================================================================
--- waagent.orig/requirements.txt
+++ waagent/requirements.txt
@@ -1,3 +1,2 @@
 distro; python_version >= '3.8'
 pyasn1
-legacycrypt; python_version >= '3.13'
Index: waagent/setup.py
===================================================================
--- waagent.orig/setup.py
+++ waagent/setup.py
@@ -324,7 +324,6 @@ class install(_install):  # pylint: disa
 #   required instead.
 requires = [
     "distro;python_version>='3.8'",
-    "legacycrypt;python_version>='3.13'",
 ]
 
 modules = []  # pylint: disable=invalid-name
