Add PexOS flavour (forked from Debian)
--- a/azurelinuxagent/agent.py
+++ b/azurelinuxagent/agent.py
@@ -196,11 +196,6 @@
         update_handler = get_update_handler()
         update_handler.run(debug)
 
-    def resourcedisk(self):
-        from azurelinuxagent.daemon.resourcedisk import get_resourcedisk_handler
-        resourcedisk_handler = get_resourcedisk_handler()
-        resourcedisk_handler.run()
-
     def show_configuration(self):
         configuration = conf.get_configuration()
         for k in sorted(configuration.keys()):
@@ -311,16 +306,16 @@
                 agent.deprovision(force, deluser=True)
             elif command == AgentCommands.Deprovision:
                 agent.deprovision(force, deluser=False)
-            elif command == AgentCommands.Provision:
-                agent.provision()
+            # elif command == AgentCommands.Provision:
+            #     agent.provision()
             elif command == AgentCommands.RegisterService:
                 agent.register_service()
             elif command == AgentCommands.Daemon:
                 agent.daemon()
             elif command == AgentCommands.RunExthandlers:
                 agent.run_exthandlers(debug)
-            elif command == AgentCommands.Resourcedisk:
-                agent.resourcedisk()
+            # elif command == AgentCommands.Resourcedisk:
+            #     agent.resourcedisk()
             elif command == AgentCommands.ShowConfig:
                 agent.show_configuration()
             elif command == AgentCommands.CollectLogs:
@@ -375,8 +370,8 @@
             cmd = AgentCommands.RegisterService
         elif re.match(regex_cmd_format.format(AgentCommands.RunExthandlers), arg):
             cmd = AgentCommands.RunExthandlers
-        elif re.match(regex_cmd_format.format(AgentCommands.Resourcedisk), arg):
-            cmd = AgentCommands.Resourcedisk
+        # elif re.match(regex_cmd_format.format(AgentCommands.Resourcedisk), arg):
+        #     cmd = AgentCommands.Resourcedisk
         elif re.match(regex_cmd_format.format(AgentCommands.Version), arg):
             cmd = AgentCommands.Version
         elif re.match(regex_cmd_format.format("verbose"), arg):
--- a/azurelinuxagent/common/osutil/default.py
+++ b/azurelinuxagent/common/osutil/default.py
@@ -423,9 +423,9 @@
             return
 
         if expiration is not None:
-            cmd = ["useradd", "-m", username, "-s", "/bin/bash", "-e", expiration]
+            cmd = ["useradd", "-m", username, "-e", expiration]
         else:
-            cmd = ["useradd", "-m", username, "-s", "/bin/bash"]
+            cmd = ["useradd", "-m", username]
 
         if comment is not None:
             cmd.extend(["-c", comment])
--- a/azurelinuxagent/common/osutil/factory.py
+++ b/azurelinuxagent/common/osutil/factory.py
@@ -40,6 +40,7 @@
 from .ubuntu import UbuntuOSUtil, Ubuntu12OSUtil, Ubuntu14OSUtil, \
     UbuntuSnappyOSUtil, Ubuntu16OSUtil, Ubuntu18OSUtil
 from .fedora import FedoraOSUtil
+from .pexos import PexOSUtil
 
 
 def get_osutil(distro_name=DISTRO_NAME,
@@ -152,5 +153,8 @@
     if distro_name == "fedora":
         return FedoraOSUtil()
 
+    if distro_name == "pexos":
+        return PexOSUtil()
+
     logger.warn("Unable to load distro implementation for {0}. Using default distro implementation instead.", distro_name)
     return DefaultOSUtil()
--- /dev/null
+++ b/azurelinuxagent/common/osutil/pexos.py
@@ -0,0 +1,280 @@
+import uuid
+
+from azurelinuxagent.common.osutil.debian import DebianOSUtil
+
+class PexOSUtil:
+    def __init__(self):
+        self._default = DebianOSUtil()
+        self.jit_enabled = False # ga/remoteaccess.py
+        self.service_name = self.get_service_name()
+
+    @staticmethod
+    def get_service_name():
+        return DebianOSUtil.get_service_name()
+
+    @staticmethod
+    def get_systemd_unit_file_install_path():
+        return DebianOSUtil.get_systemd_unit_file_install_path()
+
+    @staticmethod
+    def get_agent_bin_path():
+        return DebianOSUtil.get_agent_bin_path()
+
+    @staticmethod
+    def get_vm_arch():
+        return DebianOSUtil.get_vm_arch()
+
+    def get_firewall_dropped_packets(self, dst_ip=None):
+        # azurelinuxagent/ga/update.py
+        return 0
+
+    def get_firewall_will_wait(self):
+        # azurelinuxagent/ga/update.py
+        # azurelinuxagent/ga/persist_firewall_rules.py
+        # azurelinuxagent/common/protocol/metadata_server_migration_util.py
+        return ""
+
+    def remove_firewall(self, dst_ip, uid, wait):
+        # azurelinuxagent/common/protocol/metadata_server_migration_util.py
+        return False
+
+    def remove_legacy_firewall_rule(self, dst_ip):
+        # azurelinuxagent/ga/env.py
+        pass
+
+    def enable_firewall(self, dst_ip, uid):
+        # azurelinuxagent/common/protocol/metadata_server_migration_util.py
+        # azurelinuxagent/ga/env.py
+        return False
+
+    def get_firewall_list(self, wait=None):
+        # azurelinuxagent/ga/env.py
+        return ""
+
+    @property
+    def agent_conf_file_path(self):
+        return self._default.agent_conf_file_path
+
+    def get_agent_conf_file_path(self):
+        # azurelinuxagent/agent.py
+        return self._default.get_agent_conf_file_path()
+
+    def get_instance_id(self):
+        # azurelinuxagent/pa/provision/default.py
+        try:
+            with open("/etc/machine-id", "r") as machine_f:
+                return str(uuid.UUID(machine_f.read().strip()))
+        except (OSError, ValueError):
+            return ""
+
+    def get_root_username(self):
+        # azurelinuxagent/agent.py
+        return "root"
+
+    def useradd(self, username, expiration=None, comment=None):
+        # azurelinuxagent/ga/remoteaccess.py
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def chpasswd(self, username, password, crypt_id=6, salt_len=10):
+        # azurelinuxagent/ga/remoteaccess.py
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def get_users(self):
+        # azurelinuxagent/ga/remoteaccess.py
+        return []
+
+    def conf_sudoer(self, username, nopasswd=False, remove=False):
+        # azurelinuxagent/ga/remoteaccess.py
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def del_root_password(self):
+        # azurelinuxagent/pa/deprovision/default.py
+        pass
+
+    def deploy_ssh_keypair(self, username, keypair):
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def deploy_ssh_pubkey(self, username, pubkey):
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def conf_sshd(self, disable_password):
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def mount_dvd(self, *args, **kwargs):
+        # azurelinuxagent/common/protocol/util.py
+        # azurelinuxagent/daemon/scvmm.py
+        return self._default.mount_dvd(*args, **kwargs)
+
+    def umount_dvd(self, *args, **kwargs):
+        # azurelinuxagent/common/protocol/util.py
+        # azurelinuxagent/daemon/scvmm.py
+        return self._default.umount_dvd(*args, **kwargs)
+
+    def eject_dvd(self, chk_err=True):
+        # azurelinuxagent/common/protocol/util.py
+        return self._default.eject_dvd(chk_err)
+
+    def try_load_atapiix_mod(self):
+        # azurelinuxagent/daemon/scvmm.py
+        pass
+
+    def allow_dhcp_broadcast(self):
+        # azurelinuxagent/common/dhcp.py
+        pass
+
+    def remove_rules_files(self, rules_files=None):
+        # azurelinuxagent/ga/env.py
+        pass
+
+    def get_mac_addr(self):
+        # azurelinuxagent/common/dhcp.py
+        return ""
+
+    @staticmethod
+    def read_route_table():
+        # azurelinuxagent/common/dhcp.py
+        # azurelinuxagent/ga/monitor.py
+        return []
+
+    @staticmethod
+    def get_list_of_routes(route_table):
+        # azurelinuxagent/ga/monitor.py
+        return []
+
+    def get_dhcp_lease_endpoint(self):
+        # azurelinuxagent/common/dhcp.py
+        return None
+
+    def is_missing_default_route(self):
+        # azurelinuxagent/common/dhcp.py
+        return False
+
+    def get_if_name(self):
+        # azurelinuxagent/common/dhcp.py
+        return ""
+
+    def get_ip4_addr(self):
+        # azurelinuxagent/common/dhcp.py
+        return ""
+
+    def set_route_for_dhcp_broadcast(self, ifname):
+        # azurelinuxagent/common/dhcp.py
+        return ""
+
+    def remove_route_for_dhcp_broadcast(self, ifname):
+        # azurelinuxagent/common/dhcp.py
+        pass
+
+    def is_dhcp_available(self):
+        # azurelinuxagent/common/dhcp.py
+        # azurelinuxagent/common/protocol/util.py
+        return False
+
+    def is_dhcp_enabled(self):
+        # azurelinuxagent/common/dhcp.py
+        return False
+
+    def stop_dhcp_service(self):
+        # azurelinuxagent/common/dhcp.py
+        pass
+
+    def start_dhcp_service(self):
+        # azurelinuxagent/common/dhcp.py
+        pass
+
+    def start_network(self):
+        # azurelinuxagent/common/dhcp.py
+        pass
+
+    def start_agent_service(self):
+        # azurelinuxagent/agent.py
+        # setup.py
+        pass
+
+    def stop_agent_service(self):
+        # azurelinuxagent/agent.py
+        # azurelinuxagent/pa/deprovision/default.py
+        # setup.py
+        pass
+
+    def register_agent_service(self):
+        # azurelinuxagent/agent.py
+        # setup.py
+        pass
+
+    def restart_ssh_service(self):
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def route_add(self, net, mask, gateway):
+        # azurelinuxagent/common/dhcp.py
+        return ""
+
+    def get_dhcp_pid(self):
+        # azurelinuxagent/ga/env.py
+        return []
+
+    def set_hostname(self, hostname):
+        # azurelinuxagent/ga/env.py
+        # azurelinuxagent/pa/deprovision/default.py
+        pass
+
+    def set_dhcp_hostname(self, hostname):
+        # azurelinuxagent/pa/deprovision/default.py
+        pass
+
+    def publish_hostname(self, hostname, recover_nic=False):
+        # azurelinuxagent/ga/env.py
+        # azurelinuxagent/pa/provision/default.py
+        pass
+
+    def set_scsi_disks_timeout(self, timeout):
+        # azurelinuxagent/ga/env.py
+        pass
+
+    def get_hostname_record(self):
+        # azurelinuxagent/ga/env.py
+        return None
+
+    def del_account(self, username):
+        # azurelinuxagent/ga/remoteaccess.py
+        # azurelinuxagent/pa/deprovision/default.py
+        pass
+
+    def decode_customdata(self, data):
+        # azurelinuxagent/pa/provision/default.py
+        return ""
+
+    def get_total_mem(self):
+        # azurelinuxagent/ga/monitor.py
+        return self._default.get_total_mem()
+
+    def get_processor_cores(self):
+        # azurelinuxagent/common/event.py
+        # azurelinuxagent/ga/cpucontroller.py
+        return self._default.get_processor_cores()
+
+    def check_pid_alive(self, pid):
+        # azurelinuxagent/daemon/main.py
+        # azurelinuxagent/ga/env.py
+        # azurelinuxagent/ga/update.py
+        return self._default.check_pid_alive(pid)
+
+    def get_total_cpu_ticks_since_boot(self):
+        # azurelinuxagent/ga/cpucontroller.py
+        return self._default.get_total_cpu_ticks_since_boot()
+
+    @staticmethod
+    def get_used_and_available_system_memory():
+        # azurelinuxagent/ga/monitor.py
+        return DebianOSUtil.get_used_and_available_system_memory()
+
+    def get_nic_state(self, as_string=False):
+        # azurelinuxagent/ga/monitor.py
+        return "" if as_string else {}
--- a/azurelinuxagent/daemon/main.py
+++ b/azurelinuxagent/daemon/main.py
@@ -137,10 +137,6 @@
         if conf.get_resourcedisk_format():
             self.resourcedisk_handler.run()
 
-        # Always redetermine the protocol start (e.g., wireserver vs.
-        # on-premise) since a VHD can move between environments
-        self.protocol_util.clear_protocol()
-
         self.provision_handler.run()
 
         # Once we have the protocol, complete initialization of the telemetry fields
--- a/azurelinuxagent/daemon/resourcedisk/default.py
+++ b/azurelinuxagent/daemon/resourcedisk/default.py
@@ -18,7 +18,6 @@
 import os
 import re
 import stat
-import subprocess
 import sys
 import threading
 from time import sleep
@@ -32,7 +31,6 @@
 from azurelinuxagent.common.exception import ResourceDiskError
 from azurelinuxagent.common.osutil import get_osutil
 from azurelinuxagent.common.version import AGENT_NAME
-from azurelinuxagent.pa.provision.cloudinit import cloud_init_is_enabled
 
 DATALOSS_WARNING_FILE_NAME = "DATALOSS_WARNING_README.txt"
 DATA_LOSS_WARNING = """\
@@ -57,10 +55,6 @@
         disk_thread.start()
 
     def run(self):
-        if cloud_init_is_enabled():
-            logger.info('Using cloud-init for provisioning')
-            return
-
         mount_point = None
         if conf.get_resourcedisk_format():
             mount_point = self.activate_resource_disk()
@@ -95,8 +89,9 @@
             logger.error("Failed to enable swap {0}", e)
 
     def reread_partition_table(self, device):
-        shellutil.run("blockdev --rereadpt {0}".format(device),
-                      chk_err=False)
+        if shellutil.run("sfdisk -R {0}".format(device), chk_err=False):
+            shellutil.run("blockdev --rereadpt {0}".format(device),
+                          chk_err=False)
 
     def mount_resource_disk(self, mount_point):
         device = self.osutil.device_for_ide_port(1)
@@ -123,7 +118,7 @@
             raise ResourceDiskError(msg=msg, inner=ose)
 
         logger.info("Examining partition table")
-        ret = shellutil.run_get_output("blkid -o value -s PTTYPE {0}".format(device))
+        ret = shellutil.run_get_output("parted {0} print".format(device))
         if ret[0]:
             raise ResourceDiskError("Could not determine partition info for "
                                     "{0}: {1}".format(device, ret[1]))
@@ -134,9 +129,8 @@
         mkfs_string = "mkfs.{0} -{2} {1}".format(
             self.fs, partition, force_option)
 
-        if ret[1].strip() == "gpt":
+        if "gpt" in ret[1]:
             logger.info("GPT detected, finding partitions")
-            ret = shellutil.run_get_output("parted {0} print".format(device))
             parts = [x for x in ret[1].split("\n") if
                      re.match(r"^\s*[0-9]+", x)]
             logger.info("Found {0} GPT partition(s).", len(parts))
@@ -154,13 +148,21 @@
                 shellutil.run(mkfs_string)
         else:
             logger.info("GPT not detected, determining filesystem")
-            ret = shellutil.run_get_output("blkid -o value -s TYPE {0}".format(partition))
-            if ret[1].strip() == 'ntfs' and self.fs != 'ntfs':
+            ret = self.change_partition_type(
+                suppress_message=True,
+                option_str="{0} 1 -n".format(device))
+            ptype = ret[1].strip()
+            if ptype == "7" and self.fs != "ntfs":
                 logger.info("The partition is formatted with ntfs, updating "
                             "partition type to 83")
-                subprocess.call(['sfdisk', '-c', '-f', device, '1', '83'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+                self.change_partition_type(
+                    suppress_message=False,
+                    option_str="{0} 1 83".format(device))
+                self.reread_partition_table(device)
                 logger.info("Format partition [{0}]", mkfs_string)
                 shellutil.run(mkfs_string)
+            else:
+                logger.info("The partition type is {0}", ptype)
 
         mount_options = conf.get_resourcedisk_mountoptions()
         mount_string = self.get_mount_string(mount_options,
@@ -215,6 +217,39 @@
                     self.fs)
         return mount_point
 
+    def change_partition_type(self, suppress_message, option_str):
+        """
+            use sfdisk to change partition type.
+            First try with --part-type; if fails, fall back to -c
+        """
+
+        command_to_use = '--part-type'
+        input = "sfdisk {0} {1} {2}".format(
+            command_to_use, '-f' if suppress_message else '', option_str)
+        err_code, output = shellutil.run_get_output(
+            input, chk_err=False, log_cmd=True)
+
+        # fall back to -c
+        if err_code != 0:
+            logger.info(
+                "sfdisk with --part-type failed [{0}], retrying with -c",
+                err_code)
+            command_to_use = '-c'
+            input = "sfdisk {0} {1} {2}".format(
+                command_to_use, '-f' if suppress_message else '', option_str)
+            err_code, output = shellutil.run_get_output(input, log_cmd=True)
+
+        if err_code == 0:
+            logger.info('{0} succeeded',
+                        input)
+        else:
+            logger.error('{0} failed [{1}: {2}]',
+                         input,
+                         err_code,
+                         output)
+
+        return err_code, output
+
     def get_mount_string(self, mount_options, partition, mount_point):
         if mount_options is not None:
             return 'mount -t {0} -o {1} {2} {3}'.format(
@@ -245,6 +280,20 @@
                 os.chmod(swapfile, swapfile_mode)
             return True
 
+        return False
+
+    @staticmethod
+    def check_existing_swap_file(swapfile, swaplist, size):
+        if swapfile in swaplist and os.path.isfile(swapfile) and os.path.getsize(swapfile) == size:
+            logger.info("Swap already enabled")
+            # restrict access to owner (remove all access from group, others)
+            swapfile_mode = os.stat(swapfile).st_mode
+            if swapfile_mode & (stat.S_IRWXG | stat.S_IRWXO):
+                swapfile_mode = swapfile_mode & ~(stat.S_IRWXG | stat.S_IRWXO)
+                logger.info("Changing mode of {0} to {1:o}".format(swapfile, swapfile_mode))
+                os.chmod(swapfile, swapfile_mode)
+            return True
+
         return False
 
     def create_swap_space(self, mount_point, size_mb):
--- a/azurelinuxagent/ga/update.py
+++ b/azurelinuxagent/ga/update.py
@@ -215,9 +215,6 @@
         if child_args is not None:
             agent_cmd = "{0} {1}".format(agent_cmd, child_args)
 
-        env = os.environ.copy()
-        env['PYTHONDONTWRITEBYTECODE'] = '1'
-
         try:
 
             # Launch the correct Python version for python-based agents
@@ -233,7 +230,7 @@
                 cwd=agent_dir,
                 stdout=sys.stdout,
                 stderr=sys.stderr,
-                env=env)
+                env=os.environ)
 
             logger.verbose(u"Agent {0} launched with command '{1}'", agent_name, agent_cmd)
 
@@ -827,7 +824,7 @@
     def _ensure_readonly_files(self):
         for g in READONLY_FILE_GLOBS:
             for path in glob.iglob(os.path.join(conf.get_lib_dir(), g)):
-                os.chmod(path, stat.S_IRUSR)
+                os.chmod(path, stat.S_IRUSR | stat.S_IWUSR)
 
     def _ensure_cgroups_initialized(self):
         configurator = CGroupConfigurator.get_instance()
--- a/azurelinuxagent/pa/deprovision/factory.py
+++ b/azurelinuxagent/pa/deprovision/factory.py
@@ -23,6 +23,7 @@
 from .coreos import CoreOSDeprovisionHandler
 from .default import DeprovisionHandler
 from .ubuntu import UbuntuDeprovisionHandler, Ubuntu1804DeprovisionHandler
+from .pexos import PexOSDeprovisionHandler
 
 
 def get_deprovision_handler(distro_name=DISTRO_NAME, 
@@ -39,6 +40,8 @@
         return CoreOSDeprovisionHandler()
     if "Clear Linux" in distro_full_name:
         return ClearLinuxDeprovisionHandler()  # pylint: disable=E1120
+    if distro_name == "pexos":
+        return PexOSDeprovisionHandler()
 
     return DeprovisionHandler()
 
--- /dev/null
+++ b/azurelinuxagent/pa/deprovision/pexos.py
@@ -0,0 +1,9 @@
+class PexOSDeprovisionHandler(object):
+    def __init__(self):
+        pass
+
+    def run(self, force=False, deluser=False):
+        return
+
+    def run_changed_unique_id(self):
+        return
--- /dev/null
+++ b/config/pexos/waagent.conf
@@ -0,0 +1,126 @@
+#
+# Microsoft Azure Linux Agent Configuration
+#
+
+# Enable instance creation
+Provisioning.Agent=disabled
+
+# Enable extension handling. Do not disable this unless you do not need password reset,
+# backup, monitoring, or any extension handling whatsoever.
+Extensions.Enabled=n
+
+# Password authentication for root account will be unavailable.
+Provisioning.DeleteRootPassword=n
+
+# Generate fresh host key pair.
+Provisioning.RegenerateSshHostKeyPair=n
+
+# Supported values are "rsa", "dsa", "ecdsa", "ed25519", and "auto".
+# The "auto" option is supported on OpenSSH 5.9 (2011) and later.
+Provisioning.SshHostKeyPairType=auto
+
+# Monitor host name changes and publish changes via DHCP requests.
+Provisioning.MonitorHostName=n
+
+# Decode CustomData from Base64.
+Provisioning.DecodeCustomData=n
+
+# Execute CustomData after provisioning.
+Provisioning.ExecuteCustomData=n
+
+# Algorithm used by crypt when generating password hash.
+#Provisioning.PasswordCryptId=6
+
+# Length of random salt used when generating password hash.
+#Provisioning.PasswordCryptSaltLength=10
+
+# Allow reset password of sys user
+Provisioning.AllowResetSysUser=n
+
+# Format if unformatted. If 'n', resource disk will not be mounted.
+ResourceDisk.Format=n
+
+# File system on the resource disk
+# Typically ext3 or ext4. FreeBSD images should use 'ufs2' here.
+ResourceDisk.Filesystem=ext4
+
+# Mount point for the resource disk
+ResourceDisk.MountPoint=/mnt/resource
+
+# Create and use swapfile on resource disk.
+ResourceDisk.EnableSwap=n
+
+# Size of the swapfile.
+ResourceDisk.SwapSizeMB=0
+
+# Comma-seperated list of mount options. See man(8) for valid options.
+ResourceDisk.MountOptions=None
+
+# Enable verbose logging (y|n)
+Logs.Verbose=n
+
+# Is FIPS enabled
+OS.EnableFIPS=n
+
+# Root device timeout in seconds.
+#OS.RootDeviceScsiTimeout=300
+
+# If "None", the system default version is used.
+OS.OpensslPath=None
+
+# Set the SSH ClientAliveInterval
+# OS.SshClientAliveInterval=180
+
+# Set the path to SSH keys and configuration files
+OS.SshDir=/etc/ssh
+
+# If set, agent will use proxy server to access internet
+#HttpProxy.Host=None
+#HttpProxy.Port=None
+
+# Detect Scvmm environment, default is n
+# DetectScvmmEnv=n
+
+#
+# Lib.Dir=/var/lib/waagent
+
+#
+# DVD.MountPoint=/mnt/cdrom/secure
+
+#
+Pid.File=/run/waagent/waagent.pid
+
+#
+# Extension.LogDir=/var/log/azure
+
+#
+# Home.Dir=/home
+
+# Enable RDMA management and set up, should only be used in HPC images
+OS.EnableRDMA=n
+
+# Enable or disable goal state processing auto-update, default is enabled
+AutoUpdate.Enabled=n
+
+# Determine the update family, this should not be changed
+# AutoUpdate.GAFamily=Prod
+
+# Determine if the overprovisioning feature is enabled. If yes, hold extension
+# handling until inVMArtifactsProfile.OnHold is false.
+# Default is enabled
+# EnableOverProvisioning=y
+
+# Allow fallback to HTTP if HTTPS is unavailable
+# Note: Allowing HTTP (vs. HTTPS) may cause security risks
+# OS.AllowHTTP=n
+
+# Add firewall rules to protect access to Azure host node services
+# Note:
+# - The default is false to protect the state of existing VMs
+OS.EnableFirewall=n
+
+# Enforce control groups limits on the agent and extensions
+CGroups.EnforceLimits=n
+
+# CGroups which are excluded from limits, comma separated
+CGroups.Excluded=customscript,runcommand
--- a/setup.py
+++ b/setup.py
@@ -39,8 +39,9 @@
 
 
 def set_bin_files(data_files, dest, src=None):
-    pass
-
+    if src is None:
+        src = ["bin/waagent", "bin/waagent2.0"]
+    data_files.append((dest, src))
 
 
 def set_conf_files(data_files, dest="/etc", src=None):
@@ -251,6 +252,10 @@
         set_logrotate_files(data_files)
         set_udev_files(data_files)
         set_systemd_files(data_files, dest=systemd_dir_path)
+    elif name == 'pexos':
+        set_bin_files(data_files, dest=agent_bin_path)
+        set_conf_files(data_files, src=["config/pexos/waagent.conf"])
+        set_logrotate_files(data_files)
     else:
         # Use default setting
         set_bin_files(data_files, dest=agent_bin_path)
@@ -341,9 +346,6 @@
     install_requires=requires,
     cmdclass={
         'install': install
-    },
-    entry_points = {
-        'console_scripts': ['waagent=azurelinuxagent.agent:main'],
-    },
+    }
 )
 
--- a/tests/common/dhcp/test_dhcp.py
+++ b/tests/common/dhcp/test_dhcp.py
@@ -37,6 +37,7 @@
     def test_wireserver_route_exists(self):
         # setup
         dhcp_handler = dhcp.get_dhcp_handler()
+        dhcp_handler.osutil = osutil.DefaultOSUtil()
         self.assertTrue(dhcp_handler.endpoint is None)
         self.assertTrue(dhcp_handler.routes is None)
         self.assertTrue(dhcp_handler.gateway is None)
--- a/tests/ga/test_update.py
+++ b/tests/ga/test_update.py
@@ -21,6 +21,8 @@
 from datetime import datetime, timedelta
 from threading import current_thread
 
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.osutil.pexos import PexOSUtil
 from azurelinuxagent.ga.agent_update_handler import INITIAL_UPDATE_STATE_FILE
 from azurelinuxagent.ga.guestagent import GuestAgent, GuestAgentError, \
     AGENT_ERROR_FILE
@@ -454,7 +456,7 @@
         for path in test_files:
             mode = os.stat(path).st_mode
             mode &= (stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO)
-            self.assertEqual(0, mode ^ stat.S_IRUSR)
+            self.assertEqual(0, mode ^ (stat.S_IRUSR | stat.S_IWUSR))
 
     def test_ensure_readonly_leaves_unmodified(self):
         test_files = [
@@ -749,13 +751,12 @@
         self._test_run_latest(mock_time=mock_time)
         self.assertEqual(1, mock_time.sleep_interval)
 
-    @unittest.expectedFailure
     def test_run_latest_defaults_to_current(self):
         self.assertEqual(None, self.update_handler.get_latest_agent_greater_than_daemon())
 
         args, kwargs = self._test_run_latest()
 
-        self.assertEqual(args[0], [sys.executable, "-u", sys.argv[0], "-run-exthandlers"])
+        self.assertEqual(' '.join(args[0]), ' '.join([sys.executable, "-u", sys.argv[0], "-run-exthandlers"]))
         self.assertEqual(True, 'cwd' in kwargs)
         self.assertEqual(os.getcwd(), kwargs['cwd'])
 
@@ -1071,6 +1072,7 @@
         self.assertTrue([FirewallCmdDirectCommands.QueryPassThrough in cmd for cmd in executed_commands],
                         "The remaining commands should only be for querying the firewall commands")
 
+    @unittest.skipIf(isinstance(get_osutil(), PexOSUtil), "Not supported")
     def test_it_should_set_dns_tcp_iptable_if_drop_available_accept_unavailable(self):
 
         with TestOSUtil._mock_iptables() as mock_iptables:
@@ -1118,6 +1120,7 @@
                         self.assertEqual(filtered_mock_iptable_calls[2], accept_tcp_insert_rule,
                                          "The third command should add the accept rule")
 
+    @unittest.skipIf(isinstance(get_osutil(), PexOSUtil), "Not supported")
     def test_it_should_not_set_dns_tcp_iptable_if_drop_unavailable(self):
 
         with TestOSUtil._mock_iptables() as mock_iptables:
@@ -1156,6 +1159,7 @@
                         self.assertEqual(filtered_mock_iptable_calls[0], drop_check_command,
                                          "The first command should check the drop rule")
 
+    @unittest.skipIf(isinstance(get_osutil(), PexOSUtil), "Not supported")
     def test_it_should_not_set_dns_tcp_iptable_if_drop_and_accept_available(self):
 
         with TestOSUtil._mock_iptables() as mock_iptables:
--- /dev/null
+++ b/tests/common/osutil/test_pexos.py
@@ -0,0 +1,51 @@
+import inspect
+from unittest import TestCase
+
+from azurelinuxagent.common.osutil.default import DefaultOSUtil
+from azurelinuxagent.common.osutil.pexos import PexOSUtil
+
+
+class PexOSUtilTest(TestCase):
+
+    @staticmethod
+    def _gather_public_methods(obj: object) -> set[str]:
+        return {k for k, _ in inspect.getmembers(obj, inspect.isfunction) if not k.startswith("_")}
+
+    def test_all_default_methods_present(self) -> None:
+        """test that all DefaultOSUtil methods are implemented by PexOSUtil"""
+        # methods internal to the DefaultOSUtil
+        internal_methods = {
+            'is_selinux_enforcing',
+            'get_first_if',
+            'search_for_resource_disk',
+            'get_if_mac',
+            'get_mount_point',
+            'umount',
+            'restart_if',
+            'set_selinux_context',
+            'device_for_ide_port',
+            'get_endpoint_from_leases_path',
+            'check_and_recover_nic_state',
+            'unregister_agent_service',
+            'set_hostname_record',
+            'is_selinux_system',
+            'openssl_to_openssh',
+            'is_sys_user',
+            'set_block_device_timeout',
+            'is_loopback',
+            'get_primary_interface',
+            'restore_rules_files',
+            'mount',
+            'load_atapiix_mod',
+            'get_userentry',
+            'is_atapiix_mod_loaded',
+            'is_primary_interface',
+            'get_dvd_device',
+            'get_instance_id',
+            'is_current_instance_id',
+            'gen_password_hash',
+        }
+        self.assertSetEqual(
+            self._gather_public_methods(PexOSUtil) | internal_methods,
+            self._gather_public_methods(DefaultOSUtil)
+        )
--- a/tests/common/test_event.py
+++ b/tests/common/test_event.py
@@ -415,6 +415,7 @@
         self.assertEqual(len(event_list), 1)
         self.assertEqual(TestEvent._get_event_message(event_list[0]), u'World\u05e2\u05d9\u05d5\u05ea \u05d0\u05d7\u05e8\u05d5\u05ea\u0906\u091c')
 
+    @skip_if_predicate_true(lambda: os.getuid() == 0, "Fails when running as root")
     @skip_if_predicate_true(is_python_version_26_or_34, "Disabled on Python 2.6 and 3.4, they run on containers where the OS commands needed by the test are not present.")
     def test_collect_events_should_ignore_invalid_event_files(self):
         self._create_test_event_file("custom_script_1.tld")  # a valid event
--- a/tests/ga/test_cgroupconfigurator.py
+++ b/tests/ga/test_cgroupconfigurator.py
@@ -25,6 +25,7 @@
 import tempfile
 import time
 import threading
+from pathlib import Path
 
 from azurelinuxagent.common import conf
 from azurelinuxagent.ga.cgroupcontroller import AGENT_NAME_TELEMETRY, MetricsCounter, MetricValue, MetricsCategory
@@ -592,6 +593,7 @@
 
                         self.assertEqual(len(CGroupsTelemetry._tracked), 0, "No cgroups should have been created")
 
+    @skip_if_predicate_true(lambda: "systemd" not in Path("/proc/1/cmdline").read_text(), "Systemd not pid 1")
     @skip_if_predicate_true(is_python_version_26_or_34, "Disabled on Python 2.6 and 3.4, they run on containers where the OS commands needed by the test are not present.")
     @patch('time.sleep', side_effect=lambda _: mock_sleep())
     def test_start_extension_command_should_capture_only_the_last_subprocess_output(self, _):
